File: src/services/ai-service.js
/**
 * AI Service - Port of ai_service.py
 * Integration with Google Gemini, OpenAI, and Anthropic for portfolio analysis.
 */

const { SYSTEM_PROMPT, analysisPrompt } = require('./analysis-prompt');

class AIService {
    constructor() {
        this._provider = null;
        this._apiKey = null;
        this._model = null;
        this._userName = '';
        this.chatSessions = {}; // session_id -> chat history
        this.restoredMessages = {};
    }

    async configure(provider, apiKey, userName = '') {
        this._provider = provider;
        this._apiKey = apiKey;
        this._userName = userName;
        this._model = null;
        this.chatSessions = {}; // Reset sessions on reconfigure

        if (!apiKey) return;

        try {
            if (provider === 'gemini') {
                const { GoogleGenerativeAI } = require('@google/generative-ai');
                const genAI = new GoogleGenerativeAI(apiKey);
                this._model = genAI.getGenerativeModel({
                    model: 'gemini-2.0-flash',
                    systemInstruction: SYSTEM_PROMPT
                });
            } else if (provider === 'openai') {
                const OpenAI = require('openai');
                this._model = new OpenAI({ apiKey, timeout: 60000, maxRetries: 1 });
            } else if (provider === 'anthropic') {
                const Anthropic = require('@anthropic-ai/sdk');
                this._model = new Anthropic({ apiKey });
            }
        } catch (e) {
            console.error(`Error configuring AI (${provider}): ${e}`);
            this._model = null;
        }
    }

    isConfigured() {
        return !!(this._apiKey && this._model);
    }

    _buildSystemPrompt(strategyText = '') {
        let prompt = SYSTEM_PROMPT;
        if (this._userName) {
            prompt += `\n\nO nome do investidor é **${this._userName}**. Chame-o pelo nome.`;
        }
        if (strategyText) {
            prompt += `\n\n## ESTRATÉGIA ATUAL DO INVESTIDOR (memória persistente):\n${strategyText}\n\nEsta memória foi inferida e pode estar desatualizada. Use como hipótese contextual; priorize os dados e pedidos atuais.`;
        }
        return prompt;
    }

    _buildPortfolioContext(portfolioData) {
        if (!portfolioData) return 'Nenhum dado de portfólio disponível.';
        if (portfolioData.snapshot?.schema === 'vsa-analysis-v1') return JSON.stringify(portfolioData.snapshot, null, 2);

        const lines = ['## DADOS DO PORTFÓLIO DO INVESTIDOR\n'];

        // Assets summary
        if (portfolioData.categories) {
            lines.push('### Ativos por Categoria:');
            for (const [catName, catData] of Object.entries(portfolioData.categories)) {
                const total = catData.total || 0;
                const ativos = catData.ativos || {};
                lines.push(`\n**${catName}** (Total: R$ ${total.toLocaleString('pt-BR', { minimumFractionDigits: 2 })}):`)
                for (const [ticker, info] of Object.entries(ativos)) {
                    const quant = info.quant || 0;
                    if (quant > 0) {
                        const avgPrice = info.avgPrice ?? (quant > 0 ? info.investedVal / quant : 0);
                        const currentPrice = info.currentPrice || 0;
                        const totalVal = info.totalValue || quant * currentPrice;
                        lines.push(`  - ${ticker}: ${quant} cotas | PM: R$ ${avgPrice.toFixed(2)} | Preço Atual: R$ ${currentPrice.toFixed(2)} | Total: R$ ${totalVal.toLocaleString('pt-BR', { minimumFractionDigits: 2 })}`);
                    }
                }
            }
        }

        // Performance
        if (portfolioData.performance) {
            const perf = portfolioData.performance;
            lines.push('\n### Performance:');
            if (perf.patrimonio !== undefined) lines.push(`- Patrimônio Total: R$ ${(perf.patrimonio || 0).toLocaleString('pt-BR', { minimumFractionDigits: 2 })}`);
            if (perf.investido !== undefined) lines.push(`- Valor Investido: R$ ${(perf.investido || 0).toLocaleString('pt-BR', { minimumFractionDigits: 2 })}`);
            if (perf.lucro !== undefined) lines.push(`- Lucro/Prejuízo: R$ ${(perf.lucro || 0).toLocaleString('pt-BR', { minimumFractionDigits: 2 })}`);
            if (perf.proventos !== undefined) lines.push(`- Proventos Recebidos (período informado): R$ ${(perf.proventos || 0).toLocaleString('pt-BR', { minimumFractionDigits: 2 })}`);
            if (perf.proventos_media !== undefined) lines.push(`- Média Mensal de Proventos: R$ ${(perf.proventos_media || 0).toLocaleString('pt-BR', { minimumFractionDigits: 2 })}`);
        }

        // Rentabilidade
        if (portfolioData.rentabilidade) {
            const rent = portfolioData.rentabilidade;
            lines.push('\n### Rentabilidade estimada (Modified Dietz mensal):');
            if (rent.total !== undefined) lines.push(`- Total Acumulada: ${(rent.total || 0).toFixed(2)}%`);
            if (rent['12m'] !== undefined) lines.push(`- Últimos 12 Meses: ${(rent['12m'] || 0).toFixed(2)}%`);
            if (rent['1m'] !== undefined) lines.push(`- Último Mês: ${(rent['1m'] || 0).toFixed(2)}%`);
            if (rent.vs_cdi) lines.push(`- vs CDI: ${rent.vs_cdi}`);
        }

        // Metas
        if (portfolioData.metas) {
            lines.push('\n### Metas do Investidor:');
            for (const meta of portfolioData.metas) {
                const perc = meta.value_target > 0 ? (meta.value_current / meta.value_target * 100) : 0;
                const status = meta.kind === 'recurring' || meta.id === 'aporte_mensal' ? 'Compromisso recorrente ativo; não concluído' : perc >= 100 ? 'Objetivo atingido' : `${perc.toFixed(1)}%`;
                lines.push(`  - ${meta.title}: Atual R$ ${(meta.value_current || 0).toLocaleString('pt-BR', { minimumFractionDigits: 2 })} / Objetivo R$ ${(meta.value_target || 0).toLocaleString('pt-BR', { minimumFractionDigits: 2 })} (${status})`);
            }
        }

        // Monthly investments
        if (portfolioData.aportes) {
            const ap = portfolioData.aportes;
            lines.push('\n### Aportes:');
            lines.push(`- Aporte médio mensal: R$ ${(ap.media_mensal || 0).toLocaleString('pt-BR', { minimumFractionDigits: 2 })}`);
            lines.push(`- Total aportado no ano: R$ ${(ap.total_ano || 0).toLocaleString('pt-BR', { minimumFractionDigits: 2 })}`);
        }

        return lines.join('\n');
    }

    // ==========================================
    // GEMINI
    // ==========================================
    async _geminiAnalyze(context, strategyText = '', analysisType = 'geral') {
        const { GoogleGenerativeAI } = require('@google/generative-ai');
        const genAI = new GoogleGenerativeAI(this._apiKey);
        const model = genAI.getGenerativeModel({
            model: 'gemini-2.0-flash',
            systemInstruction: this._buildSystemPrompt(strategyText)
        });
        const prompt = analysisPrompt(context, analysisType);
        const response = await model.generateContent(prompt);
        return response.response.text();
    }

    async _geminiChat(sessionId, message, portfolioData = null, strategyText = '') {
        if (!this.chatSessions[sessionId]) {
            const context = portfolioData ? this._buildPortfolioContext(portfolioData) : '';
            const history = [];
            if (context) {
                history.push(
                    { role: 'user', parts: [{ text: `Aqui estão os dados do meu portfólio para contexto:\n\n${context}\n\nPor favor, leve esses dados em consideração em todas as respostas.` }] },
                    { role: 'model', parts: [{ text: 'Entendido! Analisei todos os dados do seu portfólio e conheço sua estratégia. Estou pronto para ajudar com sugestões personalizadas. 🚀' }] }
                );
            }
            // For Gemini, system instruction is set on the model, not the session
            // We re-create the model with strategy context for this session
            const { GoogleGenerativeAI } = require('@google/generative-ai');
            const genAI = new GoogleGenerativeAI(this._apiKey);
            const sessionModel = genAI.getGenerativeModel({
                model: 'gemini-2.0-flash',
                systemInstruction: this._buildSystemPrompt(strategyText)
            });
            history.push(...(this.restoredMessages[sessionId] || []).map(m => ({role: m.role === 'user' ? 'user' : 'model', parts: [{text: m.content}]})));
            this.chatSessions[sessionId] = sessionModel.startChat({ history });
        }

        const chat = this.chatSessions[sessionId];
        const response = await chat.sendMessage(message);
        return response.response.text();
    }

    async _geminiInferStrategy(context) {
        const { GoogleGenerativeAI } = require('@google/generative-ai');
        const genAI = new GoogleGenerativeAI(this._apiKey);
        const model = genAI.getGenerativeModel({ model: 'gemini-2.0-flash' });
        const prompt = `Com base nos dados de portfólio abaixo, infira em 3-5 frases qual é a estratégia de investimento atual deste investidor. Seja específico, use os dados reais (ativos, metas, proventos). Escreva na terceira pessoa de forma objetiva, como uma memória estratégica para um assistente de IA. Não inclua disclaimers. Responda APENAS com o texto da estratégia, sem títulos.\n\n${context}`;
        const response = await model.generateContent(prompt);
        return response.response.text().trim();
    }

    // ==========================================
    // OPENAI (ChatGPT)
    // ==========================================
    async _openaiAnalyze(context, strategyText = '', analysisType = 'geral') {
        const prompt = analysisPrompt(context, analysisType);
        const response = await this._model.chat.completions.create({
            model: 'gpt-4o-mini',
            messages: [
                { role: 'system', content: this._buildSystemPrompt(strategyText) },
                { role: 'user', content: prompt }
            ],
            temperature: 0.7,
            max_tokens: 4096
        });
        return response.choices[0].message.content;
    }

    async _openaiChat(sessionId, message, portfolioData = null, strategyText = '') {
        if (!this.chatSessions[sessionId]) {
            const context = portfolioData ? this._buildPortfolioContext(portfolioData) : '';
            const history = [{ role: 'system', content: this._buildSystemPrompt(strategyText) }];
            if (context) {
                history.push(
                    { role: 'user', content: `Aqui estão os dados do meu portfólio para contexto:\n\n${context}\n\nPor favor, leve esses dados em consideração em todas as respostas.` },
                    { role: 'assistant', content: 'Entendido! Analisei todos os dados do seu portfólio e conheço sua estratégia. Estou pronto para ajudar. 🚀' }
                );
            }
            history.push(...(this.restoredMessages[sessionId] || []).map(m => ({role: m.role === 'user' ? 'user' : 'assistant', content: m.content})));
            this.chatSessions[sessionId] = history;
        }

        this.chatSessions[sessionId].push({ role: 'user', content: message });

        const response = await this._model.chat.completions.create({
            model: 'gpt-4o-mini',
            messages: this.chatSessions[sessionId],
            temperature: 0.7,
            max_tokens: 4096
        });

        const reply = response.choices[0].message.content;
        this.chatSessions[sessionId].push({ role: 'assistant', content: reply });
        return reply;
    }

    async _openaiInferStrategy(context) {
        const response = await this._model.chat.completions.create({
            model: 'gpt-4o-mini',
            messages: [
                { role: 'system', content: 'Você é um analista financeiro que sintetiza estratégias de investimento de forma objetiva.' },
                { role: 'user', content: `Com base nos dados de portfólio abaixo, infira em 3-5 frases qual é a estratégia de investimento atual deste investidor. Seja específico, use os dados reais. Escreva na terceira pessoa de forma objetiva, como uma memória estratégica para um assistente de IA. Não inclua disclaimers. Responda APENAS com o texto da estratégia.\n\n${context}` }
            ],
            temperature: 0.5,
            max_tokens: 512
        });
        return response.choices[0].message.content.trim();
    }

    // ==========================================
    // ANTHROPIC (Claude)
    // ==========================================
    async _anthropicAnalyze(context, strategyText = '', analysisType = 'geral') {
        const prompt = analysisPrompt(context, analysisType);
        const response = await this._model.messages.create({
            model: 'claude-3-5-sonnet-latest',
            max_tokens: 4096,
            system: this._buildSystemPrompt(strategyText),
            messages: [{ role: 'user', content: prompt }]
        });
        return response.content[0].text;
    }

    async _anthropicChat(sessionId, message, portfolioData = null, strategyText = '') {
        if (!this.chatSessions[sessionId]) {
            const context = portfolioData ? this._buildPortfolioContext(portfolioData) : '';
            const history = [];
            if (context) {
                history.push(
                    { role: 'user', content: `Aqui estão os dados do meu portfólio para contexto:\n\n${context}\n\nPor favor, leve esses dados em consideração em todas as respostas.` },
                    { role: 'assistant', content: 'Entendido! Analisei todos os dados do seu portfólio e conheço sua estratégia. Estou pronto para ajudar. 🚀' }
                );
            }
            history.push(...(this.restoredMessages[sessionId] || []).map(m => ({role: m.role === 'user' ? 'user' : 'assistant', content: m.content})));
            this.chatSessions[sessionId] = history;
        }

        this.chatSessions[sessionId].push({ role: 'user', content: message });

        const response = await this._model.messages.create({
            model: 'claude-3-5-sonnet-latest',
            max_tokens: 4096,
            system: this._buildSystemPrompt(strategyText),
            messages: this.chatSessions[sessionId]
        });

        const reply = response.content[0].text;
        this.chatSessions[sessionId].push({ role: 'assistant', content: reply });
        return reply;
    }

    async _anthropicInferStrategy(context) {
        const response = await this._model.messages.create({
            model: 'claude-3-5-sonnet-latest',
            max_tokens: 512,
            system: 'Você é um analista financeiro que sintetiza estratégias de investimento de forma objetiva.',
            messages: [{ role: 'user', content: `Com base nos dados de portfólio abaixo, infira em 3-5 frases qual é a estratégia de investimento atual deste investidor. Seja específico, use os dados reais. Escreva na terceira pessoa de forma objetiva. Não inclua disclaimers. Responda APENAS com o texto da estratégia.\n\n${context}` }]
        });
        return response.content[0].text.trim();
    }

    // ==========================================
    // PUBLIC API
    // ==========================================
    async analyzePortfolio(portfolioData, strategyText = '') {
        if (!this.isConfigured()) {
            return { error: 'API key not configured. Acesse as Configurações (⚙️) para configurar.' };
        }

        const context = this._buildPortfolioContext(portfolioData);

        try {
            let text;
            const analysisType = portfolioData.analysisType || 'geral';
            
            if (this._provider === 'gemini') {
                text = await this._geminiAnalyze(context, strategyText, analysisType);
            } else if (this._provider === 'openai') {
                text = await this._openaiAnalyze(context, strategyText, analysisType);
            } else if (this._provider === 'anthropic') {
                text = await this._anthropicAnalyze(context, strategyText, analysisType);
            } else {
                return { error: `Provedor '${this._provider}' não suportado.` };
            }
            return { analysis: text };
        } catch (e) {
            const err = String(e);
            if (err.includes('429') || err.includes('Quota') || err.toLowerCase().includes('rate')) {
                return { error: 'Sem pressa! Aguarde um minuto. Você atingiu o limite de velocidade da API (Anti-Spam).' };
            }
            return { error: err };
        }
    }

    async chat(sessionId, message, portfolioData = null, strategyText = '') {
        if (!this.isConfigured()) {
            return { error: 'API key not configured. Acesse as Configurações (⚙️) para configurar.' };
        }

        try {
            let text;
            if (this._provider === 'gemini') {
                text = await this._geminiChat(sessionId, message, portfolioData, strategyText);
            } else if (this._provider === 'openai') {
                text = await this._openaiChat(sessionId, message, portfolioData, strategyText);
            } else if (this._provider === 'anthropic') {
                text = await this._anthropicChat(sessionId, message, portfolioData, strategyText);
            } else {
                return { error: `Provedor '${this._provider}' não suportado.` };
            }
            return { response: text };
        } catch (e) {
            const err = String(e);
            if (err.includes('429') || err.includes('Quota') || err.toLowerCase().includes('rate')) {
                return { error: 'Aguarde um minuto! O limite da API pediu uma pausa rápida.' };
            }
            return { error: err };
        }
    }

    /**
     * Infer the user's investment strategy using the AI.
     * Called only when the portfolio hash changes.
     */
    async inferStrategy(portfolioData, strategyText = '') {
        if (!this.isConfigured()) {
            return { error: 'API não configurada.' };
        }
        const context = this._buildPortfolioContext(portfolioData);
        try {
            let text;
            if (this._provider === 'gemini') {
                text = await this._geminiInferStrategy(context);
            } else if (this._provider === 'openai') {
                text = await this._openaiInferStrategy(context);
            } else if (this._provider === 'anthropic') {
                text = await this._anthropicInferStrategy(context);
            } else {
                return { error: `Provedor '${this._provider}' não suportado.` };
            }
            return { strategy: text };
        } catch (e) {
            const err = String(e);
            if (err.includes('429') || err.includes('Quota') || err.toLowerCase().includes('rate')) {
                return { error: 'Rate limit atingido ao inferir estratégia.' };
            }
            return { error: err };
        }
    }

    restoreSession(sessionId, messages) {
        // Rebuild each request with the current portfolio and bounded persisted history.
        delete this.chatSessions[sessionId];
        const recent = messages.slice(-40).filter(m => ['user', 'bot'].includes(m.role) && typeof m.content === 'string');
        while (recent.length && recent[0].role !== 'user') recent.shift();
        this.restoredMessages[sessionId] = recent;
    }

    clearSession(sessionId) {
        delete this.restoredMessages[sessionId];
        if (this.chatSessions[sessionId]) {
            delete this.chatSessions[sessionId];
        }
    }
}

// Singleton
const aiService = new AIService();
module.exports = { aiService, AIService };

File: ai_service.py
"""
AI Service - Integração com Google Gemini, OpenAI e Anthropic para análise de portfólio.
"""
import os
import sys

SYSTEM_PROMPT = """Você é o **InvestAI**, um consultor financeiro pessoal integrado a um dashboard de investimentos.

Seu papel é analisar o portfólio do investidor e fornecer insights personalizados, claros e acionáveis.

## Regras de conduta:
1. Sempre responda em **português brasileiro**
2. Use **emojis** para organizar e tornar a leitura agradável
3. Seja **objetivo e direto** — o investidor quer respostas práticas
4. Quando sugerir ativos específicos, sempre mencione que **não é recomendação formal de investimento**
5. Use dados concretos do portfólio nas suas análises (valores, percentuais, nomes de ativos)
6. Formate com **markdown**: use títulos, listas, negrito e itálico para organizar
7. Considere o perfil do investidor com base nos dados (conservador, moderado, arrojado)
8. Sempre relacione sugestões com as **metas** do investidor quando disponíveis
9. Se souber o nome do investidor, chame pelo nome para tornar a conversa mais pessoal

## Ao fazer análise automática, cubra:
- 📊 **Visão Geral** do portfólio
- 🎯 **Progresso das Metas** 
- ⚠️ **Pontos de Atenção** (concentração, risco, etc)
- 💡 **Sugestões** práticas e acionáveis
- 📈 **Projeções** baseadas no ritmo atual

## Disclaimer:
Sempre inclua ao final de análises mais detalhadas:
> ⚠️ *Este é um assistente educacional. As informações não constituem recomendação de investimento. Consulte um assessor certificado para decisões financeiras.*
"""


class AIService:
    def __init__(self):
        self._provider = None
        self._api_key = None
        self._model = None
        self._user_name = ""
        self.chat_sessions = {}  # session_id -> chat history

    def configure(self, provider, api_key, user_name=""):
        """Reconfigure the AI service with a new provider and API key."""
        self._provider = provider
        self._api_key = api_key
        self._user_name = user_name
        self._model = None
        self.chat_sessions = {}  # Reset sessions on reconfigure
        
        if not api_key:
            return
            
        try:
            if provider == 'gemini':
                import google.generativeai as genai
                genai.configure(api_key=api_key)
                self._model = genai.GenerativeModel(
                    model_name='gemini-2.0-flash',
                    system_instruction=SYSTEM_PROMPT
                )
            elif provider == 'openai':
                from openai import OpenAI
                self._model = OpenAI(api_key=api_key)
            elif provider == 'anthropic':
                from anthropic import Anthropic
                self._model = Anthropic(api_key=api_key)
        except Exception as e:
            print(f"Error configuring AI ({provider}): {e}")
            self._model = None

    def is_configured(self):
        return bool(self._api_key and self._model)

    def _get_system_with_name(self):
        if self._user_name:
            return SYSTEM_PROMPT + f"\n\nO nome do investidor é **{self._user_name}**. Chame-o pelo nome."
        return SYSTEM_PROMPT

    def _build_portfolio_context(self, portfolio_data):
        """Builds a text context from portfolio data for the AI."""
        if not portfolio_data:
            return "Nenhum dado de portfólio disponível."
        
        lines = ["## DADOS DO PORTFÓLIO DO INVESTIDOR\n"]
        
        # Assets summary
        if 'categories' in portfolio_data:
            lines.append("### Ativos por Categoria:")
            for cat_name, cat_data in portfolio_data['categories'].items():
                total = cat_data.get('total', 0)
                ativos = cat_data.get('ativos', {})
                lines.append(f"\n**{cat_name}** (Total: R$ {total:,.2f}):")
                for ticker, info in ativos.items():
                    quant = info.get('quant', 0)
                    if quant > 0:
                        avg_price = info.get('avgPrice', 0)
                        current_price = info.get('currentPrice', 0)
                        total_val = info.get('totalValue', quant * current_price)
                        lines.append(f"  - {ticker}: {quant} cotas | PM: R$ {avg_price:.2f} | Preço Atual: R$ {current_price:.2f} | Total: R$ {total_val:,.2f}")
        
        # Performance
        if 'performance' in portfolio_data:
            perf = portfolio_data['performance']
            lines.append(f"\n### Performance:")
            lines.append(f"- Patrimônio Total: R$ {perf.get('patrimonio', 0):,.2f}")
            lines.append(f"- Valor Investido: R$ {perf.get('investido', 0):,.2f}")
            lines.append(f"- Lucro/Prejuízo: R$ {perf.get('lucro', 0):,.2f}")
            lines.append(f"- Proventos Recebidos (12M): R$ {perf.get('proventos', 0):,.2f}")
            lines.append(f"- Média Mensal de Proventos: R$ {perf.get('proventos_media', 0):,.2f}")
        
        # Rentabilidade
        if 'rentabilidade' in portfolio_data:
            rent = portfolio_data['rentabilidade']
            lines.append(f"\n### Rentabilidade (TWR):")
            lines.append(f"- Total Acumulada: {rent.get('total', 0):.2f}%")
            lines.append(f"- Últimos 12 Meses: {rent.get('12m', 0):.2f}%")
            lines.append(f"- Último Mês: {rent.get('1m', 0):.2f}%")
            if 'vs_cdi' in rent:
                lines.append(f"- vs CDI: {rent['vs_cdi']}")
        
        # Metas
        if 'metas' in portfolio_data:
            lines.append(f"\n### Metas do Investidor:")
            for meta in portfolio_data['metas']:
                perc = (meta['value_current'] / meta['value_target'] * 100) if meta['value_target'] > 0 else 0
                status = "✅ Concluída" if perc >= 100 else f"{perc:.1f}%"
                lines.append(f"  - {meta['title']}: Atual R$ {meta['value_current']:,.2f} / Objetivo R$ {meta['value_target']:,.2f} ({status})")
        
        # Monthly investments
        if 'aportes' in portfolio_data:
            ap = portfolio_data['aportes']
            lines.append(f"\n### Aportes:")
            lines.append(f"- Aporte médio mensal: R$ {ap.get('media_mensal', 0):,.2f}")
            lines.append(f"- Total aportado no ano: R$ {ap.get('total_ano', 0):,.2f}")
        
        return "\n".join(lines)

    # ==========================================
    # GEMINI
    # ==========================================
    def _gemini_analyze(self, context):
        prompt = f"""Analise o portfólio abaixo e forneça uma análise completa e personalizada.

{context}

Forneça sua análise cobrindo: Visão Geral, Progresso das Metas, Pontos de Atenção, Sugestões práticas e Projeções."""
        
        response = self._model.generate_content(prompt)
        return response.text

    def _gemini_chat(self, session_id, message, portfolio_data=None):
        import google.generativeai as genai
        
        if session_id not in self.chat_sessions:
            context = self._build_portfolio_context(portfolio_data) if portfolio_data else ""
            history = []
            if context:
                history = [
                    {"role": "user", "parts": [f"Aqui estão os dados do meu portfólio para contexto:\n\n{context}\n\nPor favor, leve esses dados em consideração em todas as respostas."]},
                    {"role": "model", "parts": [f"Entendido! Analisei todos os dados do seu portfólio. Estou pronto para responder suas dúvidas e fornecer sugestões personalizadas com base nos seus investimentos, metas e performance. Como posso ajudar? 🚀"]}
                ]
            
            self.chat_sessions[session_id] = self._model.start_chat(history=history)
        
        chat = self.chat_sessions[session_id]
        response = chat.send_message(message)
        return response.text

    # ==========================================
    # OPENAI (ChatGPT)
    # ==========================================
    def _openai_analyze(self, context):
        prompt = f"""Analise o portfólio abaixo e forneça uma análise completa e personalizada.

{context}

Forneça sua análise cobrindo: Visão Geral, Progresso das Metas, Pontos de Atenção, Sugestões práticas e Projeções."""
        
        response = self._model.chat.completions.create(
            model="gpt-4o-mini",
            messages=[
                {"role": "system", "content": self._get_system_with_name()},
                {"role": "user", "content": prompt}
            ],
            temperature=0.7,
            max_tokens=4096
        )
        return response.choices[0].message.content

    def _openai_chat(self, session_id, message, portfolio_data=None):
        if session_id not in self.chat_sessions:
            context = self._build_portfolio_context(portfolio_data) if portfolio_data else ""
            history = [{"role": "system", "content": self._get_system_with_name()}]
            if context:
                history.append({"role": "user", "content": f"Aqui estão os dados do meu portfólio para contexto:\n\n{context}\n\nPor favor, leve esses dados em consideração em todas as respostas."})
                history.append({"role": "assistant", "content": "Entendido! Analisei todos os dados do seu portfólio. Estou pronto para responder suas dúvidas e fornecer sugestões personalizadas. Como posso ajudar? 🚀"})
            self.chat_sessions[session_id] = history
        
        self.chat_sessions[session_id].append({"role": "user", "content": message})
        
        response = self._model.chat.completions.create(
            model="gpt-4o-mini",
            messages=self.chat_sessions[session_id],
            temperature=0.7,
            max_tokens=4096
        )
        
        reply = response.choices[0].message.content
        self.chat_sessions[session_id].append({"role": "assistant", "content": reply})
        return reply

    # ==========================================
    # ANTHROPIC (Claude)
    # ==========================================
    def _anthropic_analyze(self, context):
        prompt = f"""Analise o portfólio abaixo e forneça uma análise completa e personalizada.

{context}

Forneça sua análise cobrindo: Visão Geral, Progresso das Metas, Pontos de Atenção, Sugestões práticas e Projeções."""
        
        response = self._model.messages.create(
            model="claude-3-5-sonnet-latest",
            max_tokens=4096,
            system=self._get_system_with_name(),
            messages=[
                {"role": "user", "content": prompt}
            ]
        )
        return response.content[0].text

    def _anthropic_chat(self, session_id, message, portfolio_data=None):
        if session_id not in self.chat_sessions:
            context = self._build_portfolio_context(portfolio_data) if portfolio_data else ""
            history = []
            if context:
                history.append({"role": "user", "content": f"Aqui estão os dados do meu portfólio para contexto:\n\n{context}\n\nPor favor, leve esses dados em consideração em todas as respostas."})
                history.append({"role": "assistant", "content": "Entendido! Analisei todos os dados do seu portfólio. Estou pronto para responder suas dúvidas e fornecer sugestões personalizadas. Como posso ajudar? 🚀"})
            self.chat_sessions[session_id] = history
        
        self.chat_sessions[session_id].append({"role": "user", "content": message})
        
        response = self._model.messages.create(
            model="claude-3-5-sonnet-latest",
            max_tokens=4096,
            system=self._get_system_with_name(),
            messages=self.chat_sessions[session_id]
        )
        
        reply = response.content[0].text
        self.chat_sessions[session_id].append({"role": "assistant", "content": reply})
        return reply

    # ==========================================
    # PUBLIC API
    # ==========================================
    def analyze_portfolio(self, portfolio_data):
        """Generate automatic portfolio analysis."""
        if not self.is_configured():
            return {"error": "API key not configured. Acesse as Configurações (⚙️) para configurar."}
        
        context = self._build_portfolio_context(portfolio_data)
        
        try:
            if self._provider == 'gemini':
                text = self._gemini_analyze(context)
            elif self._provider == 'openai':
                text = self._openai_analyze(context)
            elif self._provider == 'anthropic':
                text = self._anthropic_analyze(context)
            else:
                return {"error": f"Provedor '{self._provider}' não suportado."}
            
            return {"analysis": text}
        except Exception as e:
            err = str(e)
            if "429" in err or "Quota" in err or "rate" in err.lower():
                return {"error": "Sem pressa! Aguarde um minuto. Você atingiu o limite de velocidade da API (Anti-Spam)."}
            return {"error": err}

    def chat(self, session_id, message, portfolio_data=None):
        """Send a chat message with portfolio context."""
        if not self.is_configured():
            return {"error": "API key not configured. Acesse as Configurações (⚙️) para configurar."}
        
        try:
            if self._provider == 'gemini':
                text = self._gemini_chat(session_id, message, portfolio_data)
            elif self._provider == 'openai':
                text = self._openai_chat(session_id, message, portfolio_data)
            elif self._provider == 'anthropic':
                text = self._anthropic_chat(session_id, message, portfolio_data)
            else:
                return {"error": f"Provedor '{self._provider}' não suportado."}
            
            return {"response": text}
        except Exception as e:
            err = str(e)
            if "429" in err or "Quota" in err or "rate" in err.lower():
                return {"error": "Aguarde um minuto! O limite da API pediu uma pausa rápida."}
            return {"error": err}

    def clear_session(self, session_id):
        """Clear a chat session."""
        if session_id in self.chat_sessions:
            del self.chat_sessions[session_id]


# Singleton instance
ai_service = AIService()
